chia/util/condition_tools.py

```diff
@@ -55,41 +54,157 @@ def parse_sexp_to_conditions(sexp: Program) -> List[ConditionWithArgs]:
     return [parse_sexp_to_condition(s) for s in sexp.as_iter()]
 
 
+@lru_cache
+def agg_sig_additional_data(agg_sig_data: bytes) -> Dict[ConditionOpcode, bytes]:
+    ret: Dict[ConditionOpcode, bytes] = {}
+    for code in [
+        ConditionOpcode.AGG_SIG_PARENT,
+        ConditionOpcode.AGG_SIG_PUZZLE,
+        ConditionOpcode.AGG_SIG_AMOUNT,
+        ConditionOpcode.AGG_SIG_PUZZLE_AMOUNT,
+        ConditionOpcode.AGG_SIG_PARENT_AMOUNT,
+        ConditionOpcode.AGG_SIG_PARENT_PUZZLE,
+    ]:
+        ret[code] = std_hash(agg_sig_data + code)
+
+    ret[ConditionOpcode.AGG_SIG_ME] = agg_sig_data
+    return ret
+
+
 def pkm_pairs(conditions: SpendBundleConditions, additional_data: bytes) -> Tuple[List[bytes48], List[bytes]]:
     ret: Tuple[List[bytes48], List[bytes]] = ([], [])
 
+    data = agg_sig_additional_data(additional_data)
+
     for pk, msg in conditions.agg_sig_unsafe:
         ret[0].append(bytes48(pk))
         ret[1].append(msg)
-        if msg.endswith(additional_data):
-            raise ConsensusError(Err.INVALID_CONDITION)
+        for disallowed in data.values():
+            if msg.endswith(disallowed):
+                raise ConsensusError(Err.INVALID_CONDITION)
 
     for spend in conditions.spends:
+        for pk, msg in spend.agg_sig_parent:
+            ret[0].append(bytes48(pk))
```

Comment:
This is another loop that could simply use `COIN_TO_BLOB_F_LOOKUP` (see below) and iterate over `ConditionOpcode`.

Essentially, each `ConditionOpcode` has a corresponding function that takes the message from the condition, the coin, and the network identifier, and it produces a final message. Maybe we even make a function like that and use it here and [below](https://github.com/Chia-Network/chia-blockchain/pull/15769/files#diff-c9e77e30ba4a90734cb2edfde146fe37f02fa71593bf2f0b62bf045b2d68a57bR178).

```
def final_message(opcode: ConditionOpcode, spend: Coin | Spend, network_identifier: bytes32) -> bytes:
    # etc.
```

It's a bit ugly because in some places you want to pass in a `Coin` and in others a `Spend`.

Oh wait; does `Spend` even have `coin_amount`?

Reply:
I sent ac43b34 to that effect.